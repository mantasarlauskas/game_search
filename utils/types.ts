import { NextPageContext } from 'next';

export interface Info {
    id: number;
    name: string;
    url?: string;
}

export interface Game {
    name: string;
    slug: string;
    website?: string;
    background_image?: string;
    released?: string;
    platforms?: { platform: Info }[];
    genres: Info[];
    rating: number;
    ratings: { title: string, count: number }[];
    description?: string;
    background_image_additional?: string;
    ratings_count: number;
    metacritic?: number;
    developers: Info[];
    publishers: Info[];
    tags: Info[];
    esrb_rating?: Info;
    clip?: { clip: string };
}

export interface Movie {
    data: {
        480: string,
        max: string,
    }
}

export interface Screenshot {
    image: string;
}

export interface NextPageContextWithID extends NextPageContext {
    params: { id: string }
}

export interface Category {
    id: number;
    name: string;
    slug: string;
    image_background: string;
    games_count: number;
    games: { name: string; slug: string; }[];
}

export interface CategoriesPageProps {
    categories: Category[];
    nextPage?: string;
}

export interface CategoryPageProps {
    id: string;
    games: Game[];
    count: number;
    nextPage?: string;
}

export interface CategoryPageWithNameProps extends CategoryPageProps {
    name: string;
}
