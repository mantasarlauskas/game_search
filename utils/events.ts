import { MouseEvent, KeyboardEvent } from 'react';

export type ElementEvent = MouseEvent | KeyboardEvent;

export function onEnter(fn?: (e: ElementEvent) => void) {
    return (e: ElementEvent) => {
        if ((e as KeyboardEvent).key === 'Enter') {
            fn?.(e);
        }
    };
}
