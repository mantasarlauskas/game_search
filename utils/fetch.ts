export enum API_PATH {
    GAMES = 'games',
}

export async function fetchData(path: string, query?: string) {
    return fetch(
        `https://api.rawg.io/api/${path}?${query || ''}`,
    ).then((res) => {
        if (!res.ok) {
            return null;
        }

        return res.json();
    });
}
