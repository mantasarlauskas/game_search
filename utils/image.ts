export function cropImageUrl(imageUrl?: string) {
    if (!imageUrl) {
        return undefined;
    }

    const gamesIndex = imageUrl.indexOf('/games');
    const index = gamesIndex > 0 ? gamesIndex : imageUrl.indexOf('/screenshots');
    return [imageUrl.slice(0, index), '/crop/600/400', imageUrl.slice(index)].join('');
}
