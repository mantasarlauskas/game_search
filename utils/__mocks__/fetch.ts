import { getGames } from 'testing/testFactories';

module.exports = {
    fetchData: jest.fn().mockResolvedValue({
        results: getGames(),
        count: 100,
        next: '2',
        name: 'name',
    }),
    ApiPath: {
        GAMES: 'games',
        GENRES: 'genres',
        PLATFORMS: 'platforms',
        TAGS: 'tags',
        DEVELOPERS: 'developers',
        PUBLISHERS: 'publishers',
        STORES: 'stores',
    },
};
