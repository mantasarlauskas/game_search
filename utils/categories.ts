import { API_PATH, fetchData, QueryParams } from 'utils/fetch';
import { Route } from 'utils/routes';
import { Category } from 'utils/types';

export function getCategoryTitle(categories: Category[], id: string) {
    return categories.find((category) => category.id === parseInt(id, 10))?.name;
}

export async function getCategoriesServerSideProps(id: string, pageSize: number, query: QueryParams) {
    const data = await fetchData(API_PATH.GAMES, {
        page: 1,
        page_size: pageSize,
        ...query,
    });

    if (!data?.count || !data?.results) {
        return {
            redirect: {
                destination: Route.NOT_FOUND,
                permanent: false,
            },
        };
    }

    return {
        props: {
            id,
            count: data.count,
            games: data.results,
        },
    };
}

export async function getCategoryAndGamesServerSideProps(
    path: API_PATH,
    id: string,
    pageSize: number,
    query: QueryParams,
) {
    const [data, category] = await Promise.all([
        fetchData(API_PATH.GAMES, {
            page: 1,
            page_size: pageSize,
            ...query,
        }),
        fetchData(`${path}/${id}`),
    ]);

    if (!data?.count || !data?.results || !category?.name) {
        return {
            redirect: {
                destination: Route.NOT_FOUND,
                permanent: false,
            },
        };
    }

    return {
        props: {
            id,
            count: data.count,
            games: data.results,
            name: category.name,
        },
    };
}
