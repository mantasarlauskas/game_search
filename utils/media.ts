import { Movie } from 'utils/types';

export function getCroppedImageUrl(imageUrl?: string) {
    if (!imageUrl) {
        return undefined;
    }

    const gamesIndex = imageUrl.indexOf('/games');
    const index = gamesIndex > 0 ? gamesIndex : imageUrl.indexOf('/screenshots');
    return [imageUrl.slice(0, index), '/crop/600/400', imageUrl.slice(index)].join('');
}

export function getMovieUrl(movies: Movie[]) {
    return movies[Math.floor(Math.random() * movies.length)]?.data?.[480] || null;
}
