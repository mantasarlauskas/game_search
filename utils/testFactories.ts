import { Category } from 'utils/types';

export function getCategories(): Category[] {
    return [
        {
            id: 18893,
            name: 'Feral Interactive',
            slug: 'feral-interactive',
            games_count: 107,
            image_background: 'https://media.rawg.io/media/games/157/15742f2f67eacff546738e1ab5c19d20.jpg',
            games: [
                {
                    slug: 'rise-of-the-tomb-raider',
                    name: 'Rise of the Tomb Raider',
                },
                {
                    slug: 'shadow-of-mordor',
                    name: 'Middle-earth: Shadow of Mordor',
                },
                {
                    slug: 'hitman',
                    name: 'Hitman',
                },
                {
                    slug: 'bioshock-2',
                    name: 'BioShock 2',
                },
                {
                    slug: 'bioshock-remastered',
                    name: 'BioShock Remastered',
                },
                {
                    slug: 'xcom-enemy-unknown',
                    name: 'XCOM: Enemy Unknown',
                },
            ],
        },
        {
            id: 405,
            name: 'Ubisoft',
            slug: 'ubisoft',
            games_count: 297,
            image_background: 'https://media.rawg.io/media/games/5e4/5e4bff02098b2b6f824c68286d5da1a6.jpg',
            games: [
                {
                    slug: 'far-cry-3',
                    name: 'Far Cry 3',
                },
                {
                    slug: 'for-honor',
                    name: 'For Honor',
                },
                {
                    slug: 'watch-dogs',
                    name: 'Watch Dogs',
                },
                {
                    slug: 'far-cry-5',
                    name: 'Far Cry 5',
                },
                {
                    slug: 'assassins-creed-origins',
                    name: 'Assassin\'s Creed Origins',
                },
                {
                    slug: 'rayman-legends',
                    name: 'Rayman Legends',
                },
            ],
        },
    ];
}
