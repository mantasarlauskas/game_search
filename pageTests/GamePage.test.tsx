import { render, screen } from '@testing-library/react';
import GamePage, { getServerSideProps } from 'pages/games/[id]';
import { getGames } from 'testing/testFactories';
import { fetchData } from 'utils/fetch';
import { NextPageContextWithID } from 'utils/types';

jest.mock('hooks/usePaginatedQuery');
jest.mock('utils/fetch');
jest.mock('next/router', () => ({
    useRouter() {
        return {
            route: '/games',
        };
    },
}));

describe('<GamePage />', () => {
    const props = {
        game: getGames()[0],
        suggestedGames: [getGames()[1]],
    };

    it('matches snapshot', () => {
        const { container } = render(<GamePage {...props} />);
        expect(container).toMatchSnapshot();
    });

    it('shows content', () => {
        render(<GamePage {...props} />);
        expect(screen.getAllByText('2011-04-18').length).toEqual(2);
        expect(screen.getByText('Portal 2'));
        expect(screen.getByText('Tomb Raider (2013)'));
        expect(screen.getByText(/Shooter/));
        expect(screen.getByText('Puzzle'));
        expect(screen.getByText('Singleplayer'));
        expect(screen.getByText('Everyone 10+'));
        expect(screen.getByText('Games like Portal 2'));
    });

    describe('getServerSideProps', () => {
        it('returns server side props', async () => {
            const res = await getServerSideProps({ params: { id: '123' } } as NextPageContextWithID);
            expect(res).toEqual({
                props: {
                    game: expect.objectContaining({ results: getGames() }),
                    suggestedGames: getGames(),
                    suggestedGamesNextPage: '2',
                },
            });

            expect(fetchData).toBeCalledTimes(2);
            expect(fetchData).toBeCalledWith('games/123');
            expect(fetchData).toBeCalledWith('games/123/suggested', {
                page: 1,
                page_size: 6,
            });
        });
    });
});
