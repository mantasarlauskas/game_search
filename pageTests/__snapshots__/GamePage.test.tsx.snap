// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GamePage /> matches snapshot 1`] = `
<div>
  <div
    class="root"
  >
    <div
      class="releaseDate"
    >
      2011-04-18
    </div>
    <h1
      class="title"
    >
      Portal 2
    </h1>
    <div
      class="main"
    >
      <div
        class="info"
      >
        <div
          class="totalRating"
        >
          4.6
        </div>
        <div
          class="ratings"
        >
          <div
            class="rating"
          >
            exceptional:
            <div
              class="count"
            >
              2726
            </div>
          </div>
          <div
            class="rating"
          >
            recommended:
            <div
              class="count"
            >
              985
            </div>
          </div>
          <div
            class="rating"
          >
            meh:
            <div
              class="count"
            >
              104
            </div>
          </div>
          <div
            class="rating"
          >
            skip:
            <div
              class="count"
            >
              72
            </div>
          </div>
          <div
            class="rating"
          >
            Total ratings:
            <div
              class="count"
            >
              3854
            </div>
          </div>
        </div>
      </div>
      <video
        autoplay=""
        class="video"
        controls=""
        loop=""
        playsinline=""
        src="https://media.rawg.io/media/stories-640/fde/fde8aaeeab956f6b705bbb4161b09004.mp4"
      />
    </div>
    <div
      class="metaInfo"
    >
      <div
        class="root"
      >
        <div
          class="item"
        >
          <div
            class="root"
          >
            <div
              class="title"
            >
              Platforms
            </div>
            <div>
              <span
                class="item"
              >
                PC, 
              </span>
            </div>
            <div>
              <span
                class="item"
              >
                Xbox One
              </span>
            </div>
          </div>
        </div>
        <div
          class="item"
        >
          <div
            class="root"
          >
            <div
              class="title"
            >
              Metascore
            </div>
            <div
              class="score"
            >
              95
            </div>
          </div>
        </div>
        <div
          class="item"
        >
          <div
            class="root"
          >
            <div
              class="title"
            >
              Genres
            </div>
            <div>
              <span
                class="item"
              >
                Shooter, 
              </span>
            </div>
            <div>
              <span
                class="item"
              >
                Puzzle
              </span>
            </div>
          </div>
        </div>
        <div
          class="item"
        >
          <div
            class="root"
          >
            <div
              class="title"
            >
              Release date
            </div>
            2011-04-18
          </div>
        </div>
        <div
          class="item"
        >
          <div
            class="root"
          >
            <div
              class="title"
            >
              Tags
            </div>
            <div>
              <span
                class="item"
              >
                Singleplayer
              </span>
            </div>
          </div>
        </div>
        <div
          class="item"
        >
          <div
            class="root"
          >
            <div
              class="title"
            >
              Age rating
            </div>
            Everyone 10+
          </div>
        </div>
      </div>
    </div>
    <div
      class="root"
    >
      <div
        class="title"
      >
        Games like Portal 2
      </div>
      <div
        class="cards"
      >
        <div>
          <div
            class="root"
          >
            <div
              class="media"
              role="button"
              style="height: auto;"
              tabindex="0"
            >
              <img
                alt="Tomb Raider (2013)"
                class="image"
                src="https://media.rawg.io/media/crop/600/400/games/ad2/ad2ffdf80ba993654f31da045bc02456.jpg"
              />
            </div>
            <div
              class="content"
            >
              <div
                class="header"
              >
                <div
                  class="title"
                >
                  Tomb Raider (2013)
                </div>
                <div
                  class="score"
                >
                  4.1
                </div>
              </div>
              <div
                class="platforms"
              >
                Platforms: 
                <div>
                  <span
                    class="item"
                  >
                    PC, 
                  </span>
                </div>
                <div>
                  <span
                    class="item"
                  >
                    Xbox One
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="button"
      >
        <button
          class="root"
          type="button"
        >
          Load more
        </button>
      </div>
    </div>
  </div>
</div>
`;
